packages/askui-nodejs/src/core/ai-element/ai-element-collection.ts

```diff
@@ -30,42 +30,37 @@ export class AIElementCollection {
       workspaceId,
     );
 
-    if (!(await fs.pathExists(workspaceAIElementFolder))) {
-      throw new AIElementError(
-        `Missing AIElement folder for workspace '${workspaceId}' at '${workspaceAIElementFolder}'.`,
-      );
-    }
-
-    const files = await fs.readdir(workspaceAIElementFolder);
-
-    if (files.length === 0) {
-      throw new AIElementError(
-        `'${workspaceAIElementFolder}' is empty. No AIElement files found for workspace '${workspaceId}'.`,
-      );
-    }
-
-    const aiElements = await Promise.all(files
-      .filter((file) => path.extname(file) === '.json')
-      .map(async (file) => {
-        const jsonFile = path.join(workspaceAIElementFolder, file);
-        const baseName = path.basename(jsonFile, '.json');
-        const pngFile = path.join(workspaceAIElementFolder, `${baseName}.png`);
-        if (await fs.pathExists(pngFile)) {
-          const metadata: AIElementJson = JSON.parse(await fs.readFile(jsonFile, 'utf-8'));
-          return AIElement.fromJson(metadata, pngFile);
+    const sourceDirectories = [
+      workspaceAIElementFolder,
+      ...aiElementOptions.sourceDirectories.map((userPath) => path.resolve(userPath)),
+    ];
+
+    const aiElements: AIElement[] = [];
+
+    await Promise.all(
+      sourceDirectories.map(async (absoluteDirectoryName) => {
+        if (await fs.pathExists(absoluteDirectoryName)) {
+          aiElements.push(
+            ...AIElementCollection.CollectAiElementsFromDirectory(absoluteDirectoryName),
+          );
+        } else {
+          const errorMessage = `AIElements source directory '${absoluteDirectoryName}' does not exist.`;
+          if (aiElementOptions.notFoundAction === 'Warn') {
+            logger.warn(errorMessage);
+          } else if (aiElementOptions.notFoundAction === 'Error') {
+            throw new AIElementError(errorMessage);
+          }
         }
-        return null;
-      }));
-
-    const validAIElements = aiElements.filter((element): element is AIElement => element !== null);
+      }),
+    );
 
-    if (validAIElements.length === 0) {
-      throw new AIElementError(
-        `No AIElement files found for workspace '${workspaceId}' at '${workspaceAIElementFolder}'.`,
-      );
+    if (aiElements.length === 0) {
+      const formattedPaths = sourceDirectories.map((dir) => `"${dir}"`).join(', ');
+      const errorMessage = `No AIElements found in the following director${sourceDirectories.length > 1 ? 'ies' : 'y'}: [${formattedPaths}].`;
+      throw new AIElementError(errorMessage);
     }
```

Comment:
What if the user has not created an AI element (starting the lib for the first time)? I could not find it fast enough where you ensure that it does not fail then or maybe I misunderstood the logic.

Reply:
https://github.com/askui/askui/blob/ac748736f9e578603e4052ae04d575e6e3b4262d/packages/askui-nodejs/src/execution/ui-control-client.ts#L181